safe-global/safe-dao-governance-app
# refactor: improve caches + invalidation

## What it solves

Excessive RPC calls and complex cache invalidation

## How this PR fixes it

We no longer rely on `ethers`' `on` function as [it does not facilitate custom polling periods](https://docs.ethers.org/v5/concepts/events/) and our custom invalidators to reduce RPC calls:

- We instead poll the balance and contract delegate every 15 seconds when in-app in case the balance or delegate changes outside the app (this does not happen in the widgets).
- When a delegation succeeds, we also invalidate the delegate cache (set via the app itself).

All other caches (chain config, delegates file, token pause status and snapshot proposals) have been set to immutable, only changing when the address or chain ID changes ([no longer updating on window focus](https://swr.vercel.app/docs/revalidation#disable-automatic-revalidation]).

## How to test

Open the app as either a Safe App or DApp and observe no excessive log/block requests when compared to production, instead occuring every 15 seconds. They should only trigger UI changes when the balance or delegate changes.
